### ISSUE TITLE
Unusually High CPU While Fetching Mirrors

### ISSUE THREAD

**Issue by tresf** [lukesampson/scoop#1389]
I'm noticing [80-100% CPU](https://cloud.githubusercontent.com/assets/6345473/24089064/68a0f7b0-0d08-11e7-908b-2e9c8ef92712.png) utilization during the fetching process for mirrors (Server 2016 Core Eval).

Is anyone else experiencing this?  Is there an inefficiency in the CUI download indicator, or is this a bug specific to the OS/VM technology that I am on?

It doesn't seem to cause any major problems in a development environment, but for a production system, this risks the chance of stealing the resources from elsewhere.  May be all or part of the cause for the Anti-Malware executable acting up mid-install (e.g. possibly related to #1388).

**Comment by rrelmy** [comment]
I noticed that downloads are significantly slower with scoop than with a browser. I have ~10% CPU (maybe 1 of 8 HT cores?) usage on the powershell process.
Seems more than it should be. Need to investigate

**Comment by deevus** [comment]
Yeah I have noticed this as well. With the work I was doing with the
progress bars I was working towards being able to switch out the downloader
for something else if needed. Perhaps this could be user configurable.

**Comment by tresf** [comment]
I've found that [commenting out the download cursor logic](https://github.com/lukesampson/scoop/blob/master/lib/install.ps1#L167) fixes the CPU spikes introduced with the download indicator.  There are still a few other spikes, but this helps.

The cursor draw routine must have an expensive call in it although I think it can be improved.
https://github.com/lukesampson/scoop/blob/master/lib/install.ps1#L215-L252

If I'm reading this correctly:
 * The download indicator is updated with the buffer; each `2,048 B`
 * If my math is correct... `100 MB` = 48,828 executions of the draw routine.
 * Assuming we're interested in 1-100% indicator, this means on average, each percent is drawn about 500 times.

I've never done CUI interface before, so I'm not prepared to write up a PR just yet, but I think drawing the indicator less often will help remedy this problem.  There are likely other indicators that could benefit from less frequent updates.

**Comment by deevus** [comment]
Since I wrote the progress logic I'll have a look at optimising it a bit.

**Comment by tresf** [comment]
There are still some serious performance problems with the download process.

Please see right-most edge... 

With `$last = dl_onProgress $totalRead $last`
![image](https://cloud.githubusercontent.com/assets/6345473/24227939/47d93880-0f47-11e7-88be-b3a839fd0652.png)

Without `$last = dl_onProgress $totalRead $last`
![image](https://cloud.githubusercontent.com/assets/6345473/24227969/7b2511c8-0f47-11e7-8519-2e2e97251aea.png)

**Comment by deevus** [comment]
Can you try increasing the timeout to see if that helps?

**Comment by tresf** [comment]
I changed it to 1 second (1000 ms) same problem.

**Comment by deevus** [comment]
Perhaps it has to do with the number of times the event is thrown, regardless of how often we draw to the screen.

**Comment by deevus** [comment]
I'm probably not the best person to test this, because my machine is a beast (i7-5820k /w 64GB RAM)

**Comment by deevus** [comment]
What if you do call `$last = dl_onProgress $totalRead $last` but you make the function just return?

**Comment by tresf** [comment]
> Perhaps it has to do with the number of times the event is thrown, regardless of how often we draw to the screen.

I think it's a factor, but it's not a smoking gun.  Yes, this still exposes the performance problems, but so does moving the `$delta` directly into the `while {...}`.

I was able to eradicate it by using a KB counter (e.g. `byte.length * 100`), but on slow connections with very small packages (e.g. 'awk'), it doesn't really show enough progress (the timer is ideal).

Still trying things....

**Comment by tresf** [comment]
@deevus I've created a PR #1394 which seems to fix the problem.  It appears many things can spike the CPU of `powershell.exe`, namely 1. New date instantiations.  2. Frequent function calls 3. Frequent screen updates.

By using a stopwatch, it seems to fix the problem.

I blindly reverted your 01d9b95 but there is a formatting change that gets reverted.  Let me know if you need that back in.

**Comment by tresf** [comment]
For a visual... 

Before #1394 
![image](https://cloud.githubusercontent.com/assets/6345473/24230844/8ead065a-0f57-11e7-9b1b-1720af3b4508.png)

After #1394
![image](https://cloud.githubusercontent.com/assets/6345473/24230867/b574c8e0-0f57-11e7-97b0-9407795ecd51.png)

**Comment by salarmehr-cbs** [comment]
Oh this ticket is tree years old and I have exact two issues.
i installed it today, slow downloading and hight cpu usage.
